// formcomp/dist/state/form-state.svelte.js
export function createFormState(config, options = {}) {
    const { persist = 'sessionStorage', storageKey = 'formcomp-state', debounceMs = 300 } = options;
    // Initialize responses structure
    const initial = {};
    for (const step of config.steps) {
        initial[step.id] = {};
    }
    // Try to hydrate from storage
    let hydrated = initial;
    let hydratedIndex = 0;
    if (persist && typeof window !== 'undefined') {
        try {
            const storage = persist === 'sessionStorage' ? sessionStorage : localStorage;
            const stored = storage.getItem(storageKey);
            if (stored) {
                const parsed = JSON.parse(stored);
                if (parsed.responses)
                    hydrated = { ...initial, ...parsed.responses };
                if (typeof parsed.currentStepIndex === 'number')
                    hydratedIndex = parsed.currentStepIndex;
            }
        }
        catch {
            // ignore parse errors
        }
    }
    let responses = $state(hydrated);
    let currentStepIndex = $state(hydratedIndex);
    // Debounced persistence
    let saveTimer;
    function scheduleSave() {
        if (!persist || typeof window === 'undefined')
            return;
        clearTimeout(saveTimer);
        saveTimer = setTimeout(() => {
            try {
                const storage = persist === 'sessionStorage' ? sessionStorage : localStorage;
                storage.setItem(storageKey, JSON.stringify({ responses, currentStepIndex }));
            }
            catch {
                // storage full or unavailable
            }
        }, debounceMs);
    }
    return {
        get currentStepIndex() {
            return currentStepIndex;
        },
        set currentStepIndex(v) {
            currentStepIndex = v;
            scheduleSave();
        },
        get currentStepId() {
            return config.steps[currentStepIndex]?.id ?? '';
        },
        get stepCount() {
            return config.steps.length;
        },
        nextStep() {
            if (currentStepIndex < config.steps.length - 1) {
                currentStepIndex++;
                scheduleSave();
            }
        },
        prevStep() {
            if (currentStepIndex > 0) {
                currentStepIndex--;
                scheduleSave();
            }
        },
        goToStep(index) {
            if (index >= 0 && index < config.steps.length) {
                currentStepIndex = index;
                scheduleSave();
            }
        },
        getResponse(stepId, questionId) {
            return responses[stepId]?.[questionId];
        },
        setResponse(stepId, questionId, value) {
            if (!responses[stepId]) {
                responses[stepId] = {};
            }
            responses[stepId][questionId] = value;
            scheduleSave();
        },
        getStepResponses(stepId) {
            return responses[stepId] ?? {};
        },
        get allResponses() {
            return responses;
        }
    };
}
